File: javascript/ArrayList/ArrayList.js
// 迭代器
function Iterator(list) {
  this.list = list;
  this.current = 0;
}

Iterator.prototype.hasNext = function() {
  return this.current < this.list.size();
};

Iterator.prototype.next = function() {
  if (!this.hasNext()) {
    throw new Error('No Such Element Exception!');
  }

  return this.list.get(current++);
};

Iterator.prototype.remove = function() {
  this.list.remove(--current);
};


// ArrayList 构造函数
function ArrayList() {
  this.dataStore = [];
}

ArrayList.prototype.size = function() {
  return this.dataStore.length;
};

ArrayList.prototype.get = function(index) {
  if (index < 0 || index >= this.size()) {
    throw new Error('index:' + index + ' out of bounds exception!');
  }

  return this.dataStore[index];
};

ArrayList.prototype.set = function(index, newVal) {
  if (index < 0 || index >= this.size()) {
    throw new Error('index:' + index + ' out of bounds exception!');
  }

  this.dataStore[index] = newVal;
};

ArrayList.prototype.insert = function(index, val) {
  if (index < 0 || index >= this.size()) {
    throw new Error('index:' + index + ' out of bounds exception!');
  }

  this.dataStore.splice(index, 0, val);
};

ArrayList.prototype.append = function(val) {
  this.dataStore.push(val);
};

ArrayList.prototype.remove = function(index) {
  if (index < 0 || index >= this.size()) {
    throw new Error('index:' + index + ' out of bounds exception!');
  }

  this.dataStore.splice(index, 1);
};

ArrayList.prototype.clear = function() {
  this.dataStore = [];
};

ArrayList.prototype.find = function(val) {
  for (var i = 0; i < this.dataStore.length; i++) {
    if (this.dataStore[i] === val) return i;
  }
  return -1;
};

ArrayList.prototype.iterator = function() {
  return new Iterator(this);
};

ArrayList.prototype.toString = function() {
  return this.dataStore.join();
};

module.exports = ArrayList;

File: go/src/arraylist/arraylist.go
package arraylist

import (
	"errors"
	"fmt"
)

const DEFAULT_CAPACITY int = 10

type List interface {
	Size() int
	Get(index int) (interface{}, error)
	Set(index int, newVal interface{}) error
	Insert(index int, val interface{}) error
	Append(val interface{})
	Remove(index int) error
	Clear()
	String()
}

type Iterator interface {
	HasNext() bool
	Next() (interface{}, error)
	Remove()
}

type Iterable interface {
	Iterator() Iterator
}

type arrayListIterator struct {
	list    *ArrayList
	current int
}

type ArrayList struct {
	dataStore []interface{}
	theSize   int
}

func New() *ArrayList {
	list := new(ArrayList)
	list.Clear()
	return list
}

func (list *ArrayList) Size() int {
	return list.theSize
}

func (list *ArrayList) Get(index int) (interface{}, error) {
	if index < 0 || index >= list.Size() {
		return nil, errors.New("Index out of range.")
	}
	return list.dataStore[index], nil
}

func (list *ArrayList) Set(index int, newVal interface{}) error {
	if index < 0 || index >= list.Size() {
		return errors.New("Index out of range.")
	}
	list.dataStore[index] = newVal

	return nil
}

func (list *ArrayList) Append(val interface{}) {
	list.dataStore = append(list.dataStore, val)
	list.theSize++
}

func (list *ArrayList) Insert(index int, val interface{}) error {
	if index < 0 || index >= list.Size() {
		return errors.New("Index out of range.")
	}
	list.ensureCapacity()
	list.dataStore = list.dataStore[:list.Size()+1]

	for i := list.Size(); i > index; i-- {
		list.dataStore[i] = list.dataStore[i-1]
	}
	list.dataStore[index] = val
	list.theSize++

	return nil
}

func (list *ArrayList) Remove(index int) error {
	if index < 0 || index >= list.Size() {
		return errors.New("Index out of range.")
	}
	list.dataStore = append(list.dataStore[:index], list.dataStore[index+1:]...)
	list.theSize--
	return nil
}

func (list *ArrayList) Clear() {
	list.dataStore = make([]interface{}, 0, DEFAULT_CAPACITY)
	list.theSize = 0
}

func (list *ArrayList) ensureCapacity() {
	if list.Size() == cap(list.dataStore) {
		newDS := make([]interface{}, 0, 2*list.Size())
		copy(newDS, list.dataStore)
		list.dataStore = newDS
	}
}

func (list *ArrayList) String() string {
	return fmt.Sprint(list.dataStore)
}

func (list *ArrayList) Iterator() Iterator {
	iterator := new(arrayListIterator)
	iterator.current = 0
	iterator.list = list
	return iterator
}

func (it *arrayListIterator) HasNext() bool {
	return it.current < it.list.Size()
}

func (it *arrayListIterator) Next() (interface{}, error) {
	if !it.HasNext() {
		return nil, errors.New("No Such Element.")
	}
	v, err := it.list.Get(it.current)
	it.current++
	return v, err
}

func (it *arrayListIterator) Remove() {
	it.current--
	it.list.Remove(it.current)
}
